packages/storage/src/service.ts

```diff
@@ -17,55 +17,138 @@
  * @param opt_url gs:// url to a custom Storage Bucket
  *
  * @struct
  */
-export class Service {
-  authWrapper_: AuthWrapper;
-  private app_: FirebaseApp;
+export class StorageService {
+  private app_: FirebaseApp | null;
   private bucket_: Location | null = null;
   private internals_: ServiceInternals;
+  private authProvider_: Provider<FirebaseAuthInternalName>;
+  private appId_: string | null = null;
+
+  private storageRefMaker_: (loc: Location) => Reference;
+  private requestMaker_: requestMaker;
+  private pool_: XhrIoPool;
+  private requestMap_: RequestMap;
+  private deleted_: boolean = false;
+  private maxOperationRetryTime_: number;
+  private maxUploadRetryTime_: number;
 
   constructor(
-    app: FirebaseApp,
+    app: FirebaseApp | null,
     authProvider: Provider<FirebaseAuthInternalName>,
     pool: XhrIoPool,
     url?: string
   ) {
-    function maker(authWrapper: AuthWrapper, loc: Location): Reference {
-      return new Reference(authWrapper, loc);
-    }
-    this.authWrapper_ = new AuthWrapper(
-      app,
-      authProvider,
-      maker,
-      RequestExports.makeRequest,
-      this,
-      pool
-    );
     this.app_ = app;
+    this.authProvider_ = authProvider;
+    this.storageRefMaker_ = (loc: Location): Reference => {
+      return new Reference(this, loc);
+    };
+    this.requestMaker_ = makeRequest;
+    this.maxOperationRetryTime_ = constants.DEFAULT_MAX_OPERATION_RETRY_TIME;
+    this.maxUploadRetryTime_ = constants.DEFAULT_MAX_UPLOAD_RETRY_TIME;
+    this.requestMap_ = new RequestMap();
+    this.pool_ = pool;
     if (url != null) {
       this.bucket_ = Location.makeFromBucketSpec(url);
     } else {
-      const authWrapperBucket = this.authWrapper_.bucket();
-      if (authWrapperBucket != null) {
-        this.bucket_ = new Location(authWrapperBucket, '');
+      const bucketFromOptions =
+        this.app_ != null
+          ? StorageService.extractBucket_(this.app_.options)
+          : null;
+      if (bucketFromOptions != null) {
+        this.bucket_ = new Location(bucketFromOptions, '');
       }
     }
     this.internals_ = new ServiceInternals(this);
   }
 
+  private static extractBucket_(config: FirebaseOptions): string | null {
+    const bucketString = config[constants.CONFIG_STORAGE_BUCKET_KEY] || null;
+    if (bucketString == null) {
+      return null;
+    }
+    const loc: Location = Location.makeFromBucketSpec(bucketString);
+    return loc.bucket;
+  }
+
+  getAuthToken(): Promise<string | null> {
+    const auth = this.authProvider_.getImmediate({ optional: true });
+    if (auth) {
+      return auth.getToken().then(
+        (response: FirebaseAuthTokenData | null): string | null => {
+          if (response !== null) {
+            return response.accessToken;
+          } else {
+            return null;
+          }
+        },
+        () => null
+      );
+    } else {
+      return Promise.resolve(null);
+    }
+  }
```

Comment:
```suggestion
  async getAuthToken(): Promise<string | null> {
    const auth = this.authProvider_.getImmediate({ optional: true });
    if (auth) {
      const token = await; auth.getToken();
          if (token !== null) {
            return response.accessToken;
          } 
    }
    return null;
  }
```

Since we are already rewriting it :) Optional.

Reply:
Thanks, changed.